kernel/arch/dreamcast/util/vmu_pkg.c

```diff
@@ -156,3 +188,157 @@ int vmu_pkg_parse(uint8 *data, vmu_pkg_t *pkg) {
 
     return 0;
 }
+
+static unsigned int pal_get_map(uint32_t *pal, const uint32_t *curr_pal,
+                                uint8_t *map, unsigned int nb_colors) {
+    unsigned int i, j;
+
+    for (i = 0; i < 16; i++) {
+        for (j = 0; j < nb_colors; j++)
+            if (curr_pal[i] == pal[j])
+                break;
+
+        if (j < nb_colors) {
+            /* Found the color in our palette */
+            map[i] = j;
+            continue;
+        }
+
+        if (nb_colors == 15) {
+            /* No colors left :(
+             * Note that we limit to 15 colors to leave the 16th color for
+             * transparent pixels. */
+            return 0;
+        }
+
+        /* Add the new color to our palette */
+        pal[nb_colors] = curr_pal[i];
+        map[i] = nb_colors++;
+    }
+
+    return nb_colors;
+}
+
+static uint16_t argb8888_to_argb4444(uint32_t px) {
+    return 0xf000
+        | ((px >> 12) & 0x0f00)
+        | ((px >> 8) & 0x00f0)
+        | ((px >> 4) & 0x000f);
+}
+
+static void vmu_pkg_load_palette(vmu_pkg_t *pkg, const uint32_t *pal,
+                                 unsigned int nb_colors) {
+    unsigned int i;
+
+    for (i = 0; i < nb_colors; i++)
+        pkg->icon_pal[i] = argb8888_to_argb4444(pal[i]);
+
+    pkg->icon_pal[15] = 0x0; /* Transparent pixel */
+}
+
+int vmu_pkg_load_icon(vmu_pkg_t *pkg, const char *icon_fn) {
+    uint8_t px, pxh, pxl, pal_map[16];
+    unsigned int i, j, x, y, nb_colors = 0;
+    struct bmp_dib_header dib;
+    uint32_t palette[16], curr_palette[16];
+    struct ico_header hdr;
+    struct ico_dir *dir;
+    uint8_t and_mask;
+    uint8_t frame[512];
+    file_t fd;
+
+    if (!pkg->icon_cnt || !pkg->icon_data) {
+        dbglog(DBG_ERROR, "vmu_pkg_load_icon: vmu_pkg_t icon not preallocated\n");
+        return -1;
+    }
+
+    fd = fs_open(icon_fn, O_RDONLY);
+    if (fd == -1)
+        return fd;
+
+    fs_read(fd, &hdr, sizeof(hdr));
+    if (hdr.resv != 0 || hdr.type != 1) {
+        dbglog(DBG_ERROR, "vmu_pkg_load_icon: Invalid .ico header\n");
+        goto out_err_close;
+    }
+
+    /* The ICO has less frames than preallocated? No problem */
+    if (hdr.nb_images < pkg->icon_cnt)
+        pkg->icon_cnt = hdr.nb_images;
+
+    /* The ICO has more frames than preallocated? Warn and continue */
+    if (hdr.nb_images > pkg->icon_cnt) {
+        dbglog(DBG_WARNING, "vmu_pkg_load_icon: .ico file has %u frames but we only have space for %u\n",
+               hdr.nb_images, pkg->icon_cnt);
+    }
+
+    dir = alloca(pkg->icon_cnt * sizeof(*dir));
+
+    for (i = 0; i < hdr.nb_images; i++) {
+        fs_read(fd, &dir[i], sizeof(*dir));
+
+        if (dir->width != 32 || dir->height != 32 || dir->bpp != 4) {
```

Comment:
```suggestion
        if(dir->width != 32 || dir->height != 32 || dir->bpp != 4) {
```

Reply:
It might also be nice if these were all split out on their own to be more explicit as to what the error is.